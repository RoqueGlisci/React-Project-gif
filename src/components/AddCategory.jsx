// rafc

import { useState } from "react"
import PropTypes from 'prop-types'

//el setCategories entra por props desestructurado
export const AddCategory = ({onNewCategory}) => {

    const [inputValue, setinputValue] = useState('')

    //para escribir en el buscador 
    // const onInputChange = (event) => {
    //     setinputValue(event.target.value)
    // }
    // otra forma
    const onInputChange = ({target}) => {
        setinputValue(target.value)
    }

    //cundo se pulsa enter se dispara el evento
    const onSubmit = ( event ) => {
        
        event.preventDefault()
        //trim( ) elimina los espacios en blanco en ambos extremos del string. 
        //para que no se repita la enumeracion de la lista
        if(inputValue.trim().length <= 1) return
        

        //se hace un callback y desestructura para que al agregar un nuevo elemento no se reinicie los anteriores
        // setCategories(categories => [inputValue, ...categories])
        setinputValue('')//borra lo que escribimos en el buscador

        onNewCategory( inputValue.trim() );
    }

    //para escribir en el buscador
  return (
    // se maneja el formulario para inviar la informacion que hay en el inputValue
    // <form onSubmit={(event) => onSubmit(event)}> otra forma
    <form onSubmit={onSubmit} aria-label='form'>
        <input 
            type="text" 
            placeholder="Buscar gif"
            value={ inputValue }
            // onChange={ (event) => onInputChange(event) }
            // otra forma
            onChange={ onInputChange }
        />
    </form>
    
  )
}

AddCategory.propTypes = {
    onNewCategory: PropTypes.func.isRequired
}
