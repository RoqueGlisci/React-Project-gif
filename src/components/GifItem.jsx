//rafc 
import PropTypes from 'prop-types'

export const GifItem = ({title, url, id}) => {

    
  return (
    <div className="card">
        <img src={url} alt={title} />
        <p>{title}</p>
    </div>
  )
}

GifItem.propTypes = {
  title: PropTypes.string.isRequired,//isRequired es una condicion obligatoria
  url: PropTypes.string.isRequired //isRequired es una condicion obligatoria
}