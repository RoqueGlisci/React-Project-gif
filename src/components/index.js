
//archivo de barril para ordenrar las importaciones en 1 lugar. ctrl + espacio para autocompletar nombres
export * from './AddCategory'
export * from './GifGrid'
export * from './GifItem'
