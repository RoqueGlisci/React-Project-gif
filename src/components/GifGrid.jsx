//rafc 
import { PropTypes } from "prop-types";
import { GifItem } from "./GifItem"
import { useFetchGifs } from "../hooks/useFetchGifs"

export const GifGrid = ({category}) => {

    const {images, isLoading} = useFetchGifs(category)

  return (
    <>
        <h4>{category}</h4>

        {
            // si isLoading esta en falso aprarece cargando
            isLoading && <h2>Cargando...</h2>
        }

        <div className="card-grid">
            {
                /*el map busca las imagenes por el id y luego extrae el titulo, se puede desestructura en {id, title}
                se puede usar el operador (...)*/
                images.map( image => (
                    <GifItem 
                        key={image.id}
                        { ...image }
                    />
                ))
            }
        </div>

    </>
  )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}