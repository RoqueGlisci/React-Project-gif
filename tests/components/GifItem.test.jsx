import { GifItem } from "../../src/components/GifItem"
import { render, screen } from "@testing-library/react"


describe('Prueba an GifItem', () => {

    const title = 'Saitama'
    const url = 'http://one-punch.com/sitama.jpg'

    test('debe de hacer match con el snapshot', () => {
        const{container} = render(<GifItem title={title} url={url}/>)
        expect(container).toMatchSnapshot()
    })

    test('debe de mostrar la imagen con el URL y el ALT indicado', () => {
        render(<GifItem title={title} url={url}/>)

        //para no hacer 2 expect se puede desestructurar
        // expect(screen.getByRole('img').src).toBe(url)
        // expect(screen.getByRole('img').alt).toBe(title)

        //se desestructura 
        const { src, alt } = screen.getByRole('img')
        expect(src).toBe(url)
        expect(alt).toBe(alt)
    })

    test('debe de mostrar el titulo en el component', () => {
        render(<GifItem title={title} url={url}/>)
        expect(screen.getByText(title)).toBeTruthy()
    })

})