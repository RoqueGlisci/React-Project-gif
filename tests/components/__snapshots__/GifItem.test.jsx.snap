// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba an GifItem debe de hacer match con el snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Saitama"
      src="http://one-punch.com/sitama.jpg"
    />
    <p>
      Saitama
    </p>
  </div>
</div>
`;
